core/client/fs/src/main/java/alluxio/client/file/UfsFileInStream.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Supports reading from a ufs file directly.
+ */
+// TODO(lu) make it thread safe
+@NotThreadSafe
+public class UfsFileInStream extends FileInStream {
+  /** Used to manage closeable resources. */
+  private final Closer mCloser;
+  private final InputStream mUfsInStream;
+  private final long mLength;
+  private long mPosition;
+
+  /**
+   * Creates a new {@link UfsFileInStream}.
+   *
+   * @param stream the embedded input stream
+   * @param fileLength the file length
+   */
+  public UfsFileInStream(InputStream stream, long fileLength) {
+    mCloser = Closer.create();
+    mUfsInStream = Preconditions.checkNotNull(stream);
+    mLength = fileLength;
+    mPosition = 0;
+    mCloser.register(mUfsInStream);
+  }
+
+  @Override
+  public int read() throws IOException {
+    int res = mUfsInStream.read();
+    if (res > 0) {
+      mPosition += res;
+    }
+    return res;
+  }
+
+  @Override
+  public int read(ByteBuffer byteBuffer, int off, int len) throws IOException {
+    byte[] byteArray = new byte[len];
```

Comment:
there might be performance penalty for the memory allocating and copying.   I would try to make the read to fill the data to byteBuffer directly.   This might be out of scope of this PR, we could do it in a separate PR